sound/soc/sof/intel/hda-codec.c

```diff
@@ -121,12 +122,15 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
 	u32 hda_cmd = (address << 28) | (AC_NODE_ROOT << 20) |
 		(AC_VERB_PARAMETERS << 8) | AC_PAR_VENDOR_ID;
 	u32 resp = -1;
-	int ret;
+	int ret, retry = 0;
+
+	do {
+		mutex_lock(&hbus->core.cmd_mutex);
+		snd_hdac_bus_send_cmd(&hbus->core, hda_cmd);
+		snd_hdac_bus_get_response(&hbus->core, address, &resp);
```

Comment:
Hmm. Very interesting. I think this looks ok, but I wonder if we could use some of the existing retry logics. Did you try these:

You could also try:
       chip->needs_damn_long_delay = 1;

Or the bus polling mode, legacy has this in hda_controller.c:
```
static int azx_rirb_get_response(struct hdac_bus *bus, unsigned int addr,
[...]
 again:
»       err = snd_hdac_bus_get_response(bus, addr, res);
»       if (!err)
»       »       return 0;

»       if (hbus->no_response_fallback)
»       »       return -EIO;

»       if (!bus->polling_mode) {
»       »       dev_warn(chip->card->dev,
»       »       »        "azx_get_response timeout, switching to polling mode: last cmd=0x%08x\n",
»       »       »        bus->last_cmd[addr]);
»       »       bus->polling_mode = 1;
»       »       goto again;
»       }
```

Reply:
> 1. bus->needs_damn_long_delay = 1;

that's a good one :-)

Reply:
I didn't try them. I just added msleep() in different place before calling snd_hdac_bus_send_cmd(&hbus->core, hda_cmd), but it didn't help.

I will try damn_long_delay and polling_mode. will share the testing result tomorrow or next week.

thanks.

Reply:
BTW, the testing result only help us find the root cause, even damn_long_delay or polling_mode  works, we should not use it, those will add time overhead for every verb command. On those thinkstation laptops, only need a delay/retry on the first probing command.

Reply:
Thanks for testing @jason77-wang ! It may very well be that retrying the init is a sensible solution for merging, but it's a bit worrying results differ from snd-hda-intel. We've not needed the polling mode on any SOF_supported (= fairly recent) hardware, so this would the first. The damn_long_delay e.g. is used with some non-Intel HDA controllers by default. 

As for the two solutions, if polling is enabled only after one failure, then it won't slow down every verb. Similarly the damg_long_delay mode won't affect timing for the normal path. If there's a quick reply, timing is not changed.

Reply:
@kv2019i I re-read the snd_hdac_bus_get_response(), you are right, if there is a quick reply, timing is not changed. I finished the test. please see the test result in the following comment.

Reply:
After adding needs_damn_long_delay=1, the issue still could be reproduced. Below is the patch, and the dmesg_damn_log.txt is attached.
[dmesg_damn_long.txt](https://github.com/thesofproject/linux/files/7502876/dmesg_damn_long.txt)

diff --git a/sound/soc/sof/intel/hda.c b/sound/soc/sof/intel/hda.c
index f60e2c57d3d0..5a9334107ce2 100644
--- a/sound/soc/sof/intel/hda.c
+++ b/sound/soc/sof/intel/hda.c
@@ -500,6 +500,7 @@ static int hda_init(struct snd_sof_dev *sdev)
        bus->use_posbuf = 1;
        bus->bdl_pos_adj = 0;
        bus->sync_write = 1;
+       bus->needs_damn_long_delay = 1;
 
        mutex_init(&hbus->prepare_mutex);
        hbus->pci = pci;

Reply:
With the setting polling_mode, this could fix the issue on the machine, below is the patch I tried:
diff --git a/sound/soc/sof/intel/hda-codec.c b/sound/soc/sof/intel/hda-codec.c
index 6744318de612..be556f92a5fe 100644
--- a/sound/soc/sof/intel/hda-codec.c
+++ b/sound/soc/sof/intel/hda-codec.c
@@ -117,6 +117,7 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
        int type = HDA_DEV_LEGACY;
 #endif
        struct hda_bus *hbus = sof_to_hbus(sdev);
+       struct hdac_bus *bus = sof_to_bus(sdev);
        struct hdac_device *hdev;
        u32 hda_cmd = (address << 28) | (AC_NODE_ROOT << 20) |
                (AC_VERB_PARAMETERS << 8) | AC_PAR_VENDOR_ID;
@@ -125,7 +126,13 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
 
        mutex_lock(&hbus->core.cmd_mutex);
        snd_hdac_bus_send_cmd(&hbus->core, hda_cmd);
+ again:
        snd_hdac_bus_get_response(&hbus->core, address, &resp);
+       if (resp == -1 && !bus->polling_mode) {
+               bus->polling_mode = 1;
+               printk("wwwwwwwhhhhhhhhh need to set polling mode to 1\n");
+               goto again;
+       }
        mutex_unlock(&hbus->core.cmd_mutex);
        if (resp == -1)
                return -EIO;



And I found it is not the polling_mode which fixed that issue, It is just because the snd_hdac_bus_get_response() is called twice,  If I change the above the patch like below, it also could fix the issue. That is to say, when issue happens, the  snd_hdac_bus_get_response(&hbus->core, address, &resp) return the -1for resp, and if call this function one more time, it will return 10ec0287 for resp no matter setting polling_mode or not, no matter calling snd_hdac_bus_send_cmd(&hbus->core, hda_cmd) one more time or not.

Reply:
Thanks @jason77-wang for investigating all options. So the legacy driver probably handles this case via the "set polling_mode and retry" logic, although polling is strictly not needed in this case, but it will do a retry anyways and avoid the issue.

I think for the shortterm, your original patch is probably the best way to fix. I also thought about making a generic version of azx_rirb_get_response() and use that in SOF as well. But so far we've not needed the polling mode with SOF, and in this specific case, it would actually not be the best option as the controller would stay in polling mode all the time -- even if the retry was needed only at init.

Reply:
> So the legacy driver probably handles this case via the "set polling_mode and retry" logic, although polling is strictly not needed in this case, but it will do a retry anyways and avoid the issue.

@kv2019i for my education can you point me to the code that does this 'set polling mode and retry' logic?

Reply:
> > So the legacy driver probably handles this case via the "set polling_mode and retry" logic, although polling is strictly not needed in this case, but it will do a retry anyways and avoid the issue.
> 
> @kv2019i for my education can you point me to the code that does this 'set polling mode and retry' logic in the legacy driver?
@plbossart the function azx_rirb_get_response() in the sound/pci/hda/hda_controller.c

Reply:
Ah yes ok.